types/src/v0/impls/l1.rs

```diff
@@ -777,78 +866,175 @@ impl L1Client {
             }
-        });
+        }
+
         events
-            .flatten()
+            .into_iter()
             .map(|(deposit, _)| FeeInfo::from(deposit))
             .collect()
-            .await
     }
 
-    /// Get `StakeTable` at block height.
+    ////Upgrade background tasks for Proof Of Stake. No-op if upgrade
+    /// already occurred. Returns the block we need to start querying from.
+    async fn maybe_upgrade_background_tasks(&self, address: Address) {
+        if let Some(mut tasks) = self.update_task.0.lock().await.take() {
+            if tasks.len() > 1 {
+                tracing::debug!("Greater than 1 tasks are running, no need to upgrade.");
+                return;
+            } else {
+                // Protocol upgraded to POS version. If stake_update_loop is not running,
+                // we need to spawn.
+
+                tracing::warn!("Upgrading `L1Client` background tasks for v3 (Proof of Stake)!",);
+                tasks.abort_all();
+            }
+        }
+
+        let mut update_task = self.update_task.0.lock().await;
+        let mut tasks = JoinSet::new();
+        tasks.spawn(self.update_loop());
+        tasks.spawn(self.stake_update_loop(address));
+        *update_task = Some(tasks);
+
+        tracing::warn!("`Successfully upgrade L1Client background tasks!`");
+    }
+
+    /// First block we should listen to stake table events for.
+    async fn first_relevant_block_st(&self, address: Address) -> u64 {
+        let state = self.state.clone();
+        let (init_block, finalized) = {
+            let state = state.lock().await;
+            let init_block = state.stake_table_initial_block;
+            let finalized = state.snapshot.finalized.map(|block_info| block_info.number);
+            (init_block, finalized)
+        };
+
+        // if both finalized and init_block are `Some` take the
+        // maximum. If only init_block is defined, return it. We do
+        // this first to avoid unnecessary call to contract.
+        if let Some(n) = finalized
+            .zip(init_block)
+            .map(|(f, i)| f.max(i))
+            .or(init_block)
+        {
+            return n;
+        };
+
+        // Get the block contract was deployed at from the contract
+        // itself. Add it to state to avoid calling this every run. We
+        // should only hit this case in startup/early stages of
+        // run. So the next step will be to catch up stake tables from
+        // `init_block`.
+        let contract = PermissionedStakeTableInstance::new(address, self.provider.clone());
+        let init_block = if let Ok(init_block) = contract.initializedAtBlock().call().await {
+            let init_block = Some(init_block._0.to::<u64>());
+            let mut state = state.lock().await;
+            state.stake_table_initial_block = init_block;
+            init_block
+        } else {
+            tracing::error!("Failed to retrieve initial block from stake-table contract");
```

Comment:
What's the rationale for actually handling the case where we don't have the init block from the stake table? I think we can't handle this case correctly anyway because it means that the stake table contract at this address is not what we expect, or the RPC is having issues.

Reply:
I think it would be better to make this function fallible and return an error in that case.